### PR TITLE
Rydberg Drive

### PR DESCRIPTION
**Pull request by mudit2812**
Added `rydberg_drive` function to create drive Hamiltonians for neutral atom pulse programming.

Questions/notes for reviewers:
* Currently, the pulses for the `Hardwarehamiltonian` returned by `drive` (#3936) is overwritten by `rydberg_drive`. Should the `HardwarePulse` object creation be removed from `drive` and only be kept in `rydberg_drive`, and eventually `transmon_drive`?

### CODE REVIEW

File: pennylane/pulse/rydberg.py [hunk 1]
```diff
@@ -114,6 +114,147 @@ def circuit():
...
+def rydberg_drive(amplitude, phase, detuning, wires):
+    r"""Returns a :class:`ParametrizedHamiltonian` representing the action of a driving laser
+    field with the given rabi frequency, detuning and phase acting on the given wires
+
+    .. math::
+
+        \frac{1}{2} \Omega(t) \sum_{i \in \text{wires}} (\cos(\phi)\sigma_i^x - \sin(\phi)\sigma_i^y) -
+        \delta(t) \sum_{i \in \text{wires}} \sigma_i^z
+
+    where :math:`\Omega`, :math:`\delta` and :math:`\phi` correspond to the rabi frequency, detuning
```
Comment (lillian542) [inline pennylane/pulse/rydberg.py:126]:
```suggestion
    where :math:`\Omega`, :math:`\delta` and :math:`\phi` correspond to the amplitude, detuning
```

File: pennylane/pulse/rydberg.py [hunk 2]
```diff
@@ -114,6 +114,147 @@ def circuit():
...
+    and phase of the laser, :math:`i` correspond to the wire index, and :math:`\sigma^\alpha` for
```
Comment (lillian542) [inline pennylane/pulse/rydberg.py:127]:
```suggestion
    and phase of the laser, :math:`i` corresponds to the wire index, and :math:`\sigma^\alpha` for
```

File: pennylane/pulse/rydberg.py [hunk 3]
```diff
@@ -114,6 +114,147 @@ def circuit():
...
+    drive term is :math:`\mu \text{s}`. This driving term can be combined with an interaction term to
+    create a Hamiltonian describing a driven Rydberg atom system. Multiple driving terms can be combined
+    by summing them (see example).
+
+    Args:
+        amplitude (Union[float, Callable]): float or callable returning the amplitude (in MHz) of a
+            laser field
+        detuning (Union[float, Callable]): float or callable returning the detuning (in MHz) of a
+            laser field
+        phase (float): float containing the phase (in radians) of the laser field
```
Comment (lillian542) [inline pennylane/pulse/rydberg.py:138]:
```suggestion
        phase (Union[float, Callable]): float containing the phase (in radians) of the laser field
```

File: pennylane/pulse/rydberg.py [hunk 4]
```diff
@@ -114,6 +114,147 @@ def circuit():
...
+    >>> params = [2.4, [1.3, -2.0], -1.5]
+    >>> circuit_local(params)
+    Array(0.80238028, dtype=float64)
+    >>> jax.grad(circuit_local)(params)
+    [Array(-0.73273312, dtype=float64),
+     [Array(-0.45635171, dtype=float64, weak_type=True),
+      Array(0.76572817, dtype=float64, weak_type=True)],
+     Array(0.01027312, dtype=float64)]
+    """
+    if isinstance(wires, int):
```
Comment (lillian542) [inline pennylane/pulse/rydberg.py:222]: Haven't thought about this before, but are we supposed to care about custom wire labels? In theory someone could do `rydberg_drive(amp, phase, detuning, wires='a')`, right? And then I guess we still want it cast to a list?

File: pennylane/pulse/rydberg.py [hunk 5]
```diff
@@ -114,6 +114,147 @@ def circuit():
...
+    pulses = [HardwarePulse(amplitude, phase, detuning, wires)]
+
+    if amplitude_term and detuning_term:
+        drive_term = amplitude_term + detuning_term
+    elif amplitude_term:
+        drive_term = amplitude_term
+    else:
+        drive_term = detuning_term
+
+    drive_term.pulses = pulses
```
Comment (lillian542) [inline pennylane/pulse/rydberg.py:254]: 100% agree that `HardwarePulse` creation be removed from `drive` and only be kept here, no reason for it to be in `drive`

**Review by lillian542 — COMMENTED** [review]
Looking good! Left a few comments, mostly documentation suggestions

**Review by Qottmann — COMMENTED** [review]
Code-wise looks good so far!
Looking forward to checking the tests and rendered docs :)

File: pennylane/pulse/rydberg.py (hunk 4 above)
Comment (mudit2812) [reply pennylane/pulse/rydberg.py:222]: That's a good point. This should also be updated in @Qottmann 's PR.

File: pennylane/pulse/rydberg.py [hunk 6]
```diff
@@ -114,6 +115,140 @@ def circuit():
...
+
+    .. math::
+
+        \frac{1}{2} \Omega(t) \sum_{i \in \text{wires}} (\cos(\phi)\sigma_i^x - \sin(\phi)\sigma_i^y) -
+        \delta(t) \sum_{i \in \text{wires}} \sigma_i^z
+
+    where :math:`\Omega`, :math:`\delta` and :math:`\phi` correspond to the amplitude, detuning
+    and phase of the laser, :math:`i` correspond to the wire index, and :math:`\sigma^\alpha` for
+    :math:`\alpha = x,y,z` are the Pauli matrices. For hardware upload, time is expected to be in units
+    of :math:`\mu \text{s}`, and the frequency in units of :math:`MHz`. It is recommended to also follow
```
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:130]:
```suggestion
    of :math:`\text{µs}`, and the frequency in units of :math:`\text{MHz}`. It is recommended to also follow
```
Alternatively just MHz in text form (then it wouldnt be bold)

File: pennylane/pulse/rydberg.py [hunk 7]
```diff
@@ -114,6 +115,140 @@ def circuit():
     return HardwareHamiltonian(coeffs, observables, settings=settings)
 
 
+def rydberg_drive(amplitude, phase, detuning, wires):
+    r"""Returns a :class:`ParametrizedHamiltonian` representing the action of a driving laser
+    field with the given amplitude, detuning and phase acting on the given wires
```
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:120]:
```suggestion
    r"""Returns a :class:`ParametrizedHamiltonian` representing the action of a driving laser
    field,
```

File: pennylane/pulse/rydberg.py [hunk 8]
```diff
@@ -114,6 +115,140 @@ def circuit():
...
+
+        \frac{1}{2} \Omega(t) \sum_{i \in \text{wires}} (\cos(\phi)\sigma_i^x - \sin(\phi)\sigma_i^y) -
+        \delta(t) \sum_{i \in \text{wires}} \sigma_i^z
+
+    where :math:`\Omega`, :math:`\delta` and :math:`\phi` correspond to the amplitude, detuning
+    and phase of the laser, :math:`i` correspond to the wire index, and :math:`\sigma^\alpha` for
+    :math:`\alpha = x,y,z` are the Pauli matrices. For hardware upload, time is expected to be in units
+    of :math:`\mu \text{s}`, and the frequency in units of :math:`MHz`. It is recommended to also follow
+    this convention for simultation, as it avoids numerical problems due to using very large and very small
+    number. This driving term can be combined with an interaction term to create a Hamiltonian describing a
```
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:132]:
```suggestion
    numbers. This driving term can be combined with an interaction term to create a Hamiltonian describing a
```

File: pennylane/pulse/rydberg.py [hunk 9]
```diff
@@ -114,6 +115,140 @@ def circuit():
...
+    ParametrizedHamiltonian: terms=6
+    >>> H_d
+    ParametrizedHamiltonian: terms=3
+
+    The first two terms of the drive Hamiltonian ``H_d`` correspond to the first sum (the sine and cosine terms),
+    describing drive between the ground and excited states. The third term corresponds to the shift term
+    due to detuning from resonance. This drive term corresponds to a global drive that acts on all 4 wires of
+    the device.
+
+    The full Hamiltonian can be evaluated:
```
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:180]:
```suggestion
    The full Hamiltonian evolution and measuring and expectation value can be executed in a ``qnode``.
```

File: pennylane/pulse/rydberg.py [hunk 10]
```diff
@@ -114,6 +115,140 @@ def circuit():
...
+            return qml.expval(qml.PauliZ(0))
+
+    >>> params = [2.4]
+    >>> circuit(params)
+    Array(0.94301294, dtype=float64)
+    >>> jax.grad(circuit)(params)
+    [Array(0.59484969, dtype=float64)]
+
+    We can also create a Hamiltonian with multiple local drives. The following circuit corresponds to the
+    evolution where an additional local drive acting on wires ``[0, 1]`` is added to the Hamiltonian:
```
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:197]: That is a bit misleading, it could be understood that the added term consists of multiple local drives, but really it is one drive acting on wires 0 and 1 (so not fully local either)

Maybe make the example so it constructs two terms, each for qubits 0 and 1 separately (or for that manner for all qubits locally!)

File: tests/pulse/test_rydberg.py [hunk 11]
```diff
@@ -82,6 +87,125 @@ def test_max_distance(self):
...
+    def test_attributes_and_number_of_terms(self):
+        """Test that the attributes and the number of terms of the ``ParametrizedHamiltonian`` returned by
+        ``rydberg_drive`` are correct."""
+
+        Hd = rydberg_drive(amplitude=1, phase=2, detuning=3, wires=[1, 2])
+
+        assert isinstance(Hd, HardwareHamiltonian)
+        assert Hd.settings is None
+        assert Hd.wires == Wires([1, 2])
+        assert len(Hd.ops) == 3  # 2 amplitude/phase terms and one detuning term
```
Comment (Qottmann) [inline tests/pulse/test_rydberg.py:102]: remove comment

File: pennylane/pulse/rydberg.py [hunk 12]
```diff
@@ -114,6 +115,140 @@ def circuit():
...
 
 
+def rydberg_drive(amplitude, phase, detuning, wires):
+    r"""Returns a :class:`ParametrizedHamiltonian` representing the action of a driving laser
+    field with the given amplitude, detuning and phase acting on the given wires
+
+    .. math::
+
+        \frac{1}{2} \Omega(t) \sum_{i \in \text{wires}} (\cos(\phi)\sigma_i^x - \sin(\phi)\sigma_i^y) -
+        \delta(t) \sum_{i \in \text{wires}} \sigma_i^z
```
Comment (Qottmann) [inline pennylane/pulse/rydberg.py:125]:
```suggestion
        \frac{1}{2} \Omega(t) \sum_{q \in \text{wires}} (\cos(\phi)\sigma_q^x - \sin(\phi)\sigma_q^y) -
        \delta(t) \sum_{q \in \text{wires}} \sigma_q^z
```
Optional, this is to be consistent with ``drive``

**Review by Qottmann — COMMENTED** [review]
Very nice @mudit2812 some minor comments mostly concerning the docs :)

**Review by lillian542 — APPROVED** [review]
>Currently, the pulses for the Hardwarehamiltonian returned by drive (https://github.com/PennyLaneAI/pennylane/pull/3936) is overwritten by rydberg_drive. Should the HardwarePulse object creation be removed from drive and only be kept in rydberg_drive, and eventually transmon_drive?

I think the decision on the drive PR was that the `HardwarePulse` object on the drive function should be removed in this PR, and I think that's still missing and would be nice to add. Otherwise this looks good to me! 🎉

**Comment by mudit2812** [comment]
>I think the decision on the drive PR was that the HardwarePulse object on the drive function should be removed in this PR, and I think that's still missing and would be nice to add.

It's been removed 😄

**Review by Qottmann — APPROVED** [review]
Looks good to me 👍 I'd wait for the parent PR to be merged before merging this one

File: pennylane/pulse/rydberg.py [hunk 13]
```diff
@@ -114,6 +115,152 @@ def circuit():
...
+    :math:`\alpha = x,y,z` are the Pauli matrices. For hardware upload, time is expected to be in units
+    of :math:`\text{µs}`, and the frequency in units of :math:`\text{MHz}`. It is recommended to also follow
+    this convention for simultation, as it avoids numerical problems due to using very large and very small
+    numbers. This driving term can be combined with an interaction term to create a Hamiltonian describing a
+    driven Rydberg atom system. Multiple driving terms can be combined by summing them (see example).
+
+    Args:
+        amplitude (Union[float, Callable]): float or callable returning the amplitude (in MHz) of a
+            laser field
+        phase (Union[float, Callable]): float containing the phase (in radians) of the laser field
```
Comment (trbromley) [inline pennylane/pulse/rydberg.py:137]:
```suggestion
        phase (Union[float, Callable]): float or callable containing the phase (in radians) of the laser field
```

File: pennylane/pulse/rydberg.py [hunk 14]
```diff
@@ -114,6 +115,152 @@ def circuit():
...
+        detuning (Union[float, Callable]): float or callable returning the detuning (in MHz) of a
+            laser field
+        wires (Union[int, List[int]]): integer or list containing wire values for the Rydberg atoms that
+            the laser field acts on
+
+    Returns:
+        `ParametrizedHamiltonian`: representing the action of the laser field on the Rydberg atoms.
```
Comment (trbromley) [inline pennylane/pulse/rydberg.py:144]:
```suggestion
        ParametrizedHamiltonian: a :class:`~.ParametrizedHamiltonian` representing the action of the laser field on the Rydberg atoms.
```

**Review by trbromley — COMMENTED** [review]
Thanks @mudit2812!
